### PR TITLE
#159204723 Users should be able to comment on articles

### PR DESCRIPTION
**Pull request by madeofhuman**
#### What does this PR do?
- add the feature to enable users to comment on articles, reply to comments, update their comments, delete their comments, and view all comments and replies on article

#### Description of Task to be completed?
- create comment model and migrations
- add tests for the functionalities added
- create comment controller for creating a comment, getting all comments, getting a comment, updating a comment, and deleting a comment
- one-level-nest comment replies

#### How should this be manually tested?
 - clone the repo: Run `git clone https://github.com/andela/elven-ah.git` in the folder where you want the repo to be saved.
 - navigate into the newly created folder and install the dependencies: `cd elven-ah && npm install`
 - copy .env.example to .env: `cp .env.example .env`
 - provision a PostgreSQL database and add the credentials to the .env file.
 - start the server using `npm start`
 - create an account and create an article or seed some articles
 - to create a comment, make a POST request to `http://localhost:3000/api/articles/:slug/comments` using postman (Note: slug refers to the article's slug)
 - to get all comments of an article, make a GET request to `http://localhost:3000/api/articles/:slug/comments` using postman (Note: slug refers to the article's slug)
 - to get a single comment of an article, make a GET request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to retrieve)
 - to updated a comment, make a PUT request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to update)
 - to delete a comment, make a DELETE request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to delete)

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#159204723](https://www.pivotaltracker.com/story/show/159204723)

#### Screenshots (if appropriate)
screen shot 2018-08-13 at 12 43 19 pm

#### Questions:
N/A

[Finishes #159204723]

### CODE REVIEW

**Review by Veraclins — CHANGES_REQUESTED** [review]
Good job here @madeofhuman. You did a really nice job.  just have a few areas you may want to look into.
1. Is there a reason you are having a describe block for each test case? I believe all the tests related to a particular operation in the same describe block.
2. Is there a particular reason you are using Object.assign on line 55? I believe comment returned on line 54 is what you are assigning to the constant on line 55, so why are you remapping? To be honest, I can only assume what you are doing here especially since this is the only place you are doing it.

**Comment by madeofhuman** [comment]
@Veraclins thanks for taking your time to review my work. I really appreciate that. To address your concerns:

1. I have an overall `describe` block for my comment tests, inside of which I created _properly defined_ describe blocks for each test case I'm running. I did it like this to make the tests are well-defined, and clean.

2. I'm receiving an array of comments from the db, which contains data I don't need, so I'm extracting the data I need from them and performing some operations on it (like formatting the date) before returning to the user.

**Review by Veraclins — COMMENTED** [review]
@madeofhuman, I am not saying that using Object.assign is wrong or that there is a problem with the code, I just wanted to understand why you are using it. Thanks for explaining that. I now get the reason you need it. Good job.

**Review by Veraclins — APPROVED** [review]

**Review by Oluwafayokemi — APPROVED** [review]
Good Job! CJ, Weldone

File: server/controllers/CommentController.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        Comment.update({
+          body: req.body.body,
+        }, {
+          returning: true,
+          where: { id: comment.id },
+        })
+          .then(([, [updatedComment]]) => {
+            res.status(200).json({
+              status: 'success',
+              data: {
```
Comment (yomigeek) [inline server/controllers/CommentController.js]: @madeofhuman Nice work! I must implore your creativity around the approach you took regarding this task, however, based on the reference on the readme, the response returned should be "comment" and not "data".

Comment (madeofhuman) [reply server/controllers/CommentController.js]: Nice catch! I'll update that. Thank you.

**Review by chukwuemekachm — COMMENTED** [review]
I'm really impressed with this @madeofhuman, well done. However please try to validate the id parameters in the URLs. This would ensure that negative values, decimals and other characters that might cause a database crash don't get to the database.
Nice job.

**Comment by madeofhuman** [comment]
I'm glad you noticed, that, @chukwuemekachm. I will take care of that in the `catch` blocks.

**Review by chukwuemekachm — APPROVED** [review]

**Pull request merged**